# Questão 2

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import math

## 2) Um jogador de futebol americano está prestes a fazer um lançamento para outro jogador de seu time. O lançador tem uma altura de 1, 82 m e o outro jogador está 18, 2 m afastado. A expressão que descreve o movimento da bola é familiar equação da física que descreve o movimento do projétil:

$$
y = xtan(\theta)-\frac{1}{2}.\frac{ {x^{2}}g }{ {v^{2}_{0}} {cos^{2}(\theta)} } + h
$$

onde iremos isolar o y para que achemos o valor que zera a função, ficando: 
$$
0 = xtan(\theta)-\frac{1}{2}.\frac{ {x^{2}}g }{ {v^{2}_{0}} {cos^{2}(\theta)} } + h - y
$$

onde x e y são as distâncias horizontal e vertical, respectivamente, g = $9.81 \frac{m}{s^{2}}$ é a aceleração da gravidade. $v_{0}$ é a velocidade inicial da bola quando deixa a mão do lançador e θ é o ângulo que a bola faz com o eixo horizontal nesse mesmo instante. Para v0 = 15.2 m/s, x = 18, 2 m, h = 1, 82 m e y = 2, 1 m, determine o ângulo θ no qual o jogador lança a bola. Observação: o estudante não precisa saber física e nem dominar a teoria, basta substituir os valores e montar a função da qual queremos encontrar os zeros.

A função em questão fica:

$$
f(\theta) = 18.2tg(\theta) - \frac{14}{cos^{2}(\theta)} - \frac{28}{100}
$$

função que queremos encontrar o 0

Por iteração podemos encontrar o valor aproximado de teta. Montaremos um arranjo com o numpy, indo de 0 até $2\pi$, com passos pequenos de 0.0001 para melhor exatidão.
Definiremos a função para podemos jogar os valores

In [173]:
thetas = np.arange(0,2*np.pi, 0.0001)

In [174]:
def funcao(t,x,v,h,y,g):
    first = x*np.tan(t)
    second = ((x**2)*g)/(2 * (v**2) * ( np.cos(t)**2))
    return first - second + h - y

In [175]:
angulos = {}
for i in thetas:
    result = funcao(t = i, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81)
    if(math.floor(result) == 0 and result > 0):
        angulos[i] = result

In [176]:
min(angulos, key=angulos.get)

0.4616

In [177]:
angulos[0.4616]

0.0010804747745178922

A partir das iterações, descobrimos que o valor do ângulo é 0.4616.

## Aproximando o valor de $\theta$ pelo método da bisseção

O método da bisseção se aproveita do teorema de Bolzano, que é descrito da seguinte forma:

Se f:[a,b] $\: \rightarrow \: \mathbb{R}$, $y = f(x)$ é uma função contínua tal que $f(a).f(b) < 0$, então $\exists \: x^{*} \in (a,b)$ tal que $f(x^{*}) = 0$

traduzindo: dentro do intervalo no intervalo [a,b], se $f(a).f(b) < 0$, existe um x dentro do intervalo que zera a função

A questão pede que encontremos um intervalo menor que 0.1 que satisfaça o que pedimos. Ou seja, a distância entre o mínimo do intervalo e o máximo do intervalo, deve ser menos que 0.1. Vamos tomar, então, os valores 0.43 e 0.5 como teste.

In [178]:
lower = 0.43
highest = 0.5

In [179]:
highest-lower

0.07

In [180]:
a = funcao(t = lower, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81)
b = funcao(t = highest, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81)

In [181]:
a*b

-0.23632394655500588

Como a * b, que são os resultados da nossa função para 0.43 e 0.5 resultaram em um número menor que 0, que é -0.2343, então dentro deste intervalo existe um número que zera nossa função.

O método da bisseção, por sua vez, utiliza o ponto médio, iterando, para chegar no resultado. logo $x_{0} = \frac{a+b}{2}$. Dentro do método da bisseção, temos a seguinte definição:
se $f(a).f(x_{0}) < 0$ então $a = x_{0}$
ou, caso $f(x_{0}).f(b) < 0$, então $b = x_{0}$

In [182]:
def bissecao(a,b, func):
    maximo = 0
    for i in range(1000):
        maximo = (a+b)/2
        if(func(t = a, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) * func(t = maximo, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) < 0):
            b = maximo
        elif(func(t = maximo, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) * func(t = b, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) < 0):
            a = maximo
            
    return maximo
            

In [183]:
aproximacao = bissecao(lower, highest,funcao)

In [184]:
aproximacao

0.46152269273644486

In [185]:
round(funcao(t = aproximacao, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81), 3)

-0.0

Percebemos então, que pela aproximação do método da bisseção, o resultado é 0.4615

## Aproximando o valor de $\theta$ por Newton-Raphson

O método de Newton-Rapshon é um pouco diferente, mas ainda é um método iterativo que utiliza derivadas.
O método de Newton-Raphson e definido da seguinte forma:

$$
x^{(n+1)} = x^{(n)} - \frac{f(x^{(n)})}{f'(x^{n})}
$$

A derivada de nossa função, ou $f'(x)$ é:

$$
f'(x) = xsec^{2}(\theta) - \frac{ x^{2} g sec^{2}(\theta) tan^{2}(\theta) }{v^{2}_{0}}
$$

In [186]:
def derivada(t,x,g,v):
    first = x * (1/(np.cos(t)**2))
    second = (x**2 * g * (1/(np.cos(t)**2))) * (np.tan(t)**2)/v**2
    return first - second

In [187]:
def newton_raphson():
    val = 0
    for i in range(1000):
        val = val - funcao(t = val, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81)/derivada(t = val, x=18.2, v=15.2, g=9.81)
    return val

In [188]:
newton_raphson()

0.46152269273644486

Pecebemos então que a resposta pelo Newton Raphson, que é 0.4615 bate com as demais.

## Use a Posição Falsa para encontrar uma solução com erro menor que $10^{-4}$.

O método da Posição Falsa é uma modificação do método da bisseção.
Como o método da bisseção converge de maneira lenta, faremos uma mudança no algorítmo.
$$
x_{n} = \frac{af(b)-bf(a)}{f(b)-f(a)}
$$
Assim como no método da bisseção, temos que ter $f(a)f(b) < 0$ para que a resposta esteja no intervalo, e temos a mesma relação:
$$
Se \: f(a)f(x_{n}) < 0 \: entao\: a = x_{n}
\\
ou \: se f(b)f(x_{n}) < 0 \: entao \: b = x_{n}
$$

In [189]:
def posicao_falsa(a,b, func):
    resultado = 0
    for i in range(1000):
        resultado = (a*func(t = b, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) - b*func(t = a, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81))/(func(t = b, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) - func(t = a, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81))
        if(func(t = a, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) * func(t = resultado, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) < 0):
            b = resultado
        elif(func(t = resultado, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) * func(t = b, x=18.2, v=15.2, h=1.82, y=2.1, g=9.81) < 0):
            a = resultado
        
    return resultado

In [190]:
posicao_falsa(0.43, 0.5, funcao)

0.46152269273644486

Percebemos, então que todos os resultados batem. Logo, nosso $\theta$ é 0.46152269273644486

In [203]:
T = 0.46152269273644486

# a) Use a bisseção para encontrar um intervalo menor que 0.1 que contenha a raiz da equação.

In [202]:
def funcao_x(t,x,v,h,g):
    first = x*np.tan(t)
    second = ((x**2)*g)/(2 * (v**2) * ( np.cos(t)**2))
    return first - second + h

In [226]:
a = funcao_x(T,x=-3.2 ,v=15.2,h=1.82,g=9.81)

In [225]:
b = funcao_x(T,x=-3.1 ,v=15.2,h=1.82,g=9.81)

In [227]:
a*b

-0.0010131516205250624

In [215]:
def bissecao_x(a,b, func):
    maximo = 0
    for i in range(1000):
        maximo = (a+b)/2
        if(func(t = T, x=a, v=15.2, h=1.82, g=9.81) * func(t = T, x=maximo, v=15.2, h=1.82, g=9.81) < 0):
            b = maximo
        elif(func(t = T, x=b, v=15.2, h=1.82, g=9.81) * func(t = T, x=maximo, v=15.2, h=1.82, g=9.81) < 0):
            a = maximo
            
    return maximo

In [230]:
bissecao_x(-3.2,-3.1,funcao_x)

-3.135833157438121

# b) Use Newton-Rapshon para encontrar uma solução com erro menor que $10^{-4}$

Derivada da função em relação a x:
$$
y = tan(\theta) - \frac{x.g}{v^{2}cos^{2}(\theta)}
$$

In [232]:
def derivada_x(t, x,v,g):
    result = np.tan(t) - (x*g)/((v**2) * (np.cos(t)**2))
    return result

In [233]:
def newton_raphson_x():
    val = 0
    for i in range(1000):
        val = val - funcao_x(T,x=val ,v=15.2,h=1.82,g=9.81)/derivada_x(t = T, x=val, v=15.2, g=9.81)
    return val

In [237]:
newton_raphson_x()

-3.1358331574381215

# C) Use a Posição Falsa para encontrar uma solução com erro menor que $10^{-4}$

In [235]:
def posicao_falsa_x(a,b, func):
    resultado = 0
    for i in range(1000):
        resultado = (a*func(T,x=b ,v=15.2,h=1.82,g=9.81) - b*func(T,x=a ,v=15.2,h=1.82,g=9.81))/(func(T,x=b ,v=15.2,h=1.82,g=9.81) - func(T,x=a ,v=15.2,h=1.82,g=9.81))
        if(func(T,x=b ,v=15.2,h=1.82,g=9.81) * func(T,x=resultado ,v=15.2,h=1.82,g=9.81) < 0):
            b = resultado
        elif(func(T,x=a ,v=15.2,h=1.82,g=9.81) * func(T,x=resultado ,v=15.2,h=1.82,g=9.81) < 0):
            a = resultado
        
    return resultado

In [238]:
posicao_falsa_x(-3.2,-3.1,funcao_x)

-3.135833157438121